Source/VMware.vSphereDSC/DSCResources/VMHostAccount.ps1

```diff
@@ -0,0 +1,279 @@
+            }
+        }
+
+        return ($null -ne $this.Description -and $this.Description -ne $vmHostAccount.Description)
+    }
+
+    <#
+    .DESCRIPTION
+
+    Populates the parameters for the New-VMHostAccount and Set-VMHostAccount cmdlets.
+    #>
+    [void] PopulateVMHostAccountParams($vmHostAccountParams, $parameter, $desiredValue) {
+        if ($null -ne $desiredValue) {
+            $vmHostAccountParams.$parameter = $desiredValue
+        }
+    }
+
+    <#
+    .DESCRIPTION
+
+    Returns the populated VMHost Account parameters.
+    #>
+    [hashtable] GetVMHostAccountParams() {
+        $vmHostAccountParams = @{}
+
+        $vmHostAccountParams.Server = $this.Connection
+        $vmHostAccountParams.Confirm = $false
+        $vmHostAccountParams.ErrorAction = 'Stop'
+
+        $this.PopulateVMHostAccountParams($vmHostAccountParams, $this.AccountPasswordParameterName, $this.AccountPassword)
+        $this.PopulateVMHostAccountParams($vmHostAccountParams, $this.DescriptionParameterName, $this.Description)
+
+        return $vmHostAccountParams
+    }
+
+    <#
+    .DESCRIPTION
+
+    Creates a new Permission with the passed Role for the specified VMHost Account.
+    #>
+    [void] CreateAccountPermission($vmHostAccount) {
+        $accountRole = Get-VIRole -Server $this.Connection -Name $this.Role -ErrorAction SilentlyContinue
+        if ($null -eq $accountRole) {
+            throw "The passed role $($this.Role) is not present on the server."
+        }
+
+        try {
+            New-VIPermission -Server $this.Connection -Entity $this.Server -Principal $vmHostAccount -Role $accountRole -ErrorAction Stop
+        }
+        catch {
+            throw "Cannot assign role $($this.Role) to account $($vmHostAccount.Name). For more information: $($_.Exception.Message)"
+        }
+    }
+
+    <#
+    .DESCRIPTION
+
+    Creates a new VMHost Account with the specified properties.
+    #>
+    [void] AddVMHostAccount() {
+        $vmHostAccountParams = $this.GetVMHostAccountParams()
+        $vmHostAccountParams.Id = $this.Id
+
+        $vmHostAccount = $null
+
+        try {
+            $vmHostAccount = New-VMHostAccount @vmHostAccountParams
+        }
+        catch {
+            throw "Cannot create VMHost Account $($this.Id). For more information: $($_.Exception.Message)"
+        }
+
+        $this.CreateAccountPermission($vmHostAccount)
+    }
+
+    <#
+    .DESCRIPTION
+
+    Updates the VMHost Account with the specified properties.
+    #>
+    [void] UpdateVMHostAccount($vmHostAccount) {
+        $vmHostAccountParams = $this.GetVMHostAccountParams()
+
+        try {
+            $vmHostAccount | Set-VMHostAccount @vmHostAccountParams
+        }
+        catch {
+            throw "Cannot update VMHost Account $($this.Id). For more information: $($_.Exception.Message)"
+        }
+
+        if ($this.ShouldCreateAcountPermission($vmHostAccount)) {
+            $this.CreateAccountPermission($vmHostAccount)
```

Comment:
Creating new permission on update operation, what happens in this case? Are you sure the new permission isn't added to the existing one. Check whether account could have multiple permissions on same entity

Reply:
Yes, one account can have only one permission on the same entity. So on update operation new permission is created and it overrides the existing one. Tested with the following script:

```
 $viServer = Connect-VIServer -Server '<server>' -User '<user>' -Password '<password>'
 New-VMHostAccount -Server $viServer -Id 'MyTestAccount' -Password '<account password>'

 $vmHostAccount = Get-VMHostAccount -Server $viServer -Name 'MyTestAccount'
 $entity = Get-VMHost -Server $viServer
 $role = Get-VIRole -Server $viServer -Name 'MyCustomRole1'

 New-VIPermission -Server $viServer -Entity $entity -Principal $vmHostAccount -Role $role

 $role = Get-VIRole -Server $viServer -Name 'MyCustomRole2'
 
 New-VIPermission -Server $viServer -Entity $entity -Principal $vmHostAccount -Role $role

 $permission = Get-VIPermission -Server $viServer -Principal $vmHostAccount
 $permission.Role // It is equal to 'MyCustomRole2'
```